fix toExponential()

yanovich opened the pull request:
Closes #28

munrocket commented:
Can you explain how you convert second number to string?
Probably some references or whatever.

yanovich commented:
> Can you explain how you convert second number to string?
> Probably some references or whatever.

I subtract already connected part until the remainder is zero or less. The meaningful part of the low number moves to the high number with each subtraction. Here is my debug version. I moved it outside to enable `console.log()`.
```
$ cat a.mjs
import * as Double from 'double.js/dist/double.js'
//import { Double } from 'double.js/dist/double.esm.js'
//require = require("esm")(module)
//const D = require('double.js').Double
const D = Double.default.Double

function toExponential(self, precision) {
  if (isNaN(self.hi)) {
    return 'NaN';
  }
  if (!isFinite(self.hi) || self.toNumber() == 0.) {
    return self.hi.toExponential(precision);
  }
  let remainder = D.clone(self);
  let str = remainder.hi.toExponential(precision).split('e');
  if (str[0].length > 17) {
    str[0] = str[0].slice(0, 17)
  }
  let result = str[0];
  console.log('. i:', str[0].length, str[0].indexOf('.'), 1)
  let i = str[0].length - str[0].indexOf('.') - 1;
  if (str[0].indexOf('.') < 0) {
    i--;
  }
  D.sub22(remainder, new D(result + 'e' + str[1]))
  if (remainder.hi < 0.) {
    D.mul21(remainder, -1.0);
  }
  if (precision !== undefined) {
    if (precision > 33) {
      precision = 33;
    }
  }
  while (true) {
    console.log('. result:', result, 'remainder', remainder, i)
    let nextPrecision = undefined;
    if (precision === undefined) {
      if (remainder.toNumber() <= 0.) {
        break;
      }
    } else {
      if (i >= precision) {
        break;
      }
      if (remainder.toNumber() <= 0.) {
        result += '0';
        i++;
        continue;
      }
      nextPrecision = precision - i;
      if (nextPrecision > 14) {
        nextPrecision = 14;
      }
    }
    let next = remainder.hi.toExponential(nextPrecision).split('e');
    let nextDigs = next[0].replace(/^0\.|\./, '');
    let nextLength = nextDigs.length;
    if (nextLength > 15) {
      nextLength = 15;
    }
    if (precision === undefined) {
      if ((nextLength + i) > 33) {
        nextLength = 33 - i;
      }
    } else {
      if ((nextLength + i) > precision) {
        nextLength = precision - i;
      }
    }
    nextDigs = nextDigs.slice(0, nextLength);
    result += nextDigs;
    i += nextLength;
    if (i >= 33) {
      break;
    }
    D.sub22(remainder, new D(nextDigs + 'e' + next[1]))
  }
  return result + 'e' + str[1];
}

console.log(D)

const inner_radius = new D('1.0')
console.log(inner_radius.toNumber())
const outer_radius = new D('2.0')
console.log(outer_radius.toNumber())
const radius = outer_radius.add(inner_radius)
console.log(radius.toNumber())
console.log(outer_radius)
console.log(inner_radius)
console.log(radius)
console.log(toExponential(radius))
console.log(toExponential(radius, 5))
console.log(radius)
console.log(toExponential(D.Pi))
console.log(toExponential(new D('3.14159265358979323846264338')))
console.log(toExponential(new D('-3.14159265358979323846264338')))
```

munrocket commented:
LGTM. Thanks!
